bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java

```diff
@@ -550,11 +599,451 @@ private static RetryOptions createRetryOptions(Configuration configuration) {
+  private static HeaderProvider buildHeaderProvider(Configuration configuration) {
+
+    // This information is in addition to bigtable-client-core version, and jdk version.
+    StringBuilder agentBuilder = new StringBuilder();
+    agentBuilder.append("hbase-").append(VersionInfo.getVersion());
+    String customUserAgent = configuration.get(CUSTOM_USER_AGENT_KEY);
+    if (customUserAgent != null) {
+      agentBuilder.append(',').append(customUserAgent);
+    }
+
+    return FixedHeaderProvider.create(
+        USER_AGENT_KEY.name(), VENEER_ADAPTER + agentBuilder.toString());
+  }
+
+  /** Creates {@link CredentialsProvider} based on {@link CredentialOptions}. */
+  private static CredentialsProvider buildCredentialProvider(Configuration configuration)
+      throws IOException {
+
+    Credentials credentials = null;
+    if (configuration.getBoolean(
+        BIGTABLE_USE_SERVICE_ACCOUNTS_KEY, BIGTABLE_USE_SERVICE_ACCOUNTS_DEFAULT)) {
+      LOG.debug("Using service accounts");
+
+      if (configuration instanceof BigtableExtendedConfiguration) {
+        credentials = ((BigtableExtendedConfiguration) configuration).getCredentials();
+
+      } else if (configuration.get(BIGTABLE_SERVICE_ACCOUNT_JSON_VALUE_KEY) != null) {
+        String jsonValue = configuration.get(BIGTABLE_SERVICE_ACCOUNT_JSON_VALUE_KEY);
+        LOG.debug("Using json value");
+
+        Preconditions.checkState(
+            !isNullOrEmpty(jsonValue), "service account json value is null or empty");
+        credentials =
+            GoogleCredentials.fromStream(
+                new ByteArrayInputStream(jsonValue.getBytes(StandardCharsets.UTF_8)));
+
+      } else if (configuration.get(BIGTABLE_SERVICE_ACCOUNT_JSON_KEYFILE_LOCATION_KEY) != null) {
+
+        String keyFileLocation =
+            configuration.get(BIGTABLE_SERVICE_ACCOUNT_JSON_KEYFILE_LOCATION_KEY);
+        LOG.debug("Using json keyfile: %s", keyFileLocation);
+
+        Preconditions.checkState(
+            !isNullOrEmpty(keyFileLocation), "service account location is null or empty");
+        credentials = GoogleCredentials.fromStream(new FileInputStream(keyFileLocation));
+      } else if (configuration.get(BIGTABLE_SERVICE_ACCOUNT_EMAIL_KEY) != null) {
+        String serviceAccount = configuration.get(BIGTABLE_SERVICE_ACCOUNT_EMAIL_KEY);
+        LOG.debug("Service account %s specified.", serviceAccount);
+
+        String keyFileLocation =
+            configuration.get(BIGTABLE_SERVICE_ACCOUNT_P12_KEYFILE_LOCATION_KEY);
+        Preconditions.checkState(
+            !isNullOrEmpty(keyFileLocation),
+            "Key file location must be specified when setting service account email");
+        LOG.debug("Using p12 keyfile: %s", keyFileLocation);
+
+        credentials = getCredentialFromPrivateKeyServiceAccount(serviceAccount, keyFileLocation);
+      } else {
+        LOG.debug("Using default credentials.");
+        credentials = GoogleCredentials.getApplicationDefault();
```

Comment:
This will break directpath. We should not override the default credentials unless the user specified something in specific